File: notion_utils.rb
def notion_client
  @notion_client ||= Notion::Client.new(token: ENV['NOTION_TOKEN'])
end

def fetch_cached_page_blocks(page_id, last_edited_time)
  cache_key = "page_#{page_id}_#{last_edited_time}_blocks"
  $cache.fetch(cache_key) do
    fetch_notion_page_blocks(page_id)
  end
end

def extract_page_id_from_notion_url(url)
  uuid_match = url.match(/-([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})$/i)
  if uuid_match
    return uuid_match[1]
  end

  hex_match = url.match(/-([a-f0-9]{32})$/i)
  if hex_match
    id = hex_match[1]
    return "#{id[0..7]}-#{id[8..11]}-#{id[12..15]}-#{id[16..19]}-#{id[20..31]}"
  end

  raise "Could not extract page ID from Notion URL: #{url}"
end

def find_database_id(content_page_id, database_name)
  blocks = []
  response = notion_client.block_children(block_id: content_page_id, page_size: 100)
  blocks.concat(response['results'] || [])

  while response['has_more'] && response['next_cursor']
    response = notion_client.block_children(
      block_id: content_page_id,
      page_size: 100,
      start_cursor: response['next_cursor']
    )
    blocks.concat(response['results'] || [])
  end

  blocks.each do |block|
    if block['type'] == 'child_database'
      title = block['child_database']['title'] || ''
      if title.downcase.include?(database_name.downcase)
        return block['id']
      end
    end
  end

  raise "#{database_name.capitalize} database not found in CONTENT_NOTION_URL page"
end

def find_chirps_database_id(content_page_id)
  find_database_id(content_page_id, 'chirp')
end

def find_posts_database_id(content_page_id)
  find_database_id(content_page_id, 'post')
end

def find_notion_page_by_title(content_page_id, page_title)
  blocks = []
  response = notion_client.block_children(block_id: content_page_id, page_size: 100)
  blocks.concat(response['results'] || [])

  while response['has_more'] && response['next_cursor']
    response = notion_client.block_children(
      block_id: content_page_id,
      page_size: 100,
      start_cursor: response['next_cursor']
    )
    blocks.concat(response['results'] || [])
  end

  blocks.each do |block|
    if block['type'] == 'child_page'
      title = block['child_page']['title'] || ''
      if title.downcase == page_title.downcase
        return block['id']
      end
    end
  end

  raise "#{page_title} page not found in CONTENT_NOTION_URL page"
end

def fetch_notion_page_by_name(page_name)
  content_notion_url = ENV['CONTENT_NOTION_URL']
  raise "CONTENT_NOTION_URL not found in environment variables" if content_notion_url.nil? || content_notion_url.empty?

  content_page_id = extract_page_id_from_notion_url(content_notion_url)
  page_id = find_notion_page_by_title(content_page_id, page_name)

  page_response = notion_client.page(page_id: page_id)
  page = page_response

  properties = page['properties'] || {}
  last_edited_time = page['last_edited_time'] || page[:last_edited_time]

  cache_key = "page_#{page_name}_#{last_edited_time}"
  $cache.fetch(cache_key) do
    title = nil
    properties.each do |prop_name, prop|
      if prop['type'] == 'title' || prop[:type] == 'title'
        title = prop['title']&.first&.[]('plain_text') if prop['title']&.any?
        break if title
      end
    end
    title ||= page_name.capitalize

    date_prop = properties['date'] || properties['Date'] || properties['created_time'] || properties['Created Time']
    date = if date_prop
      if date_prop['type'] == 'date' && date_prop['date']
        date_prop['date']['start']
      elsif date_prop['type'] == 'created_time'
        date_prop['created_time']
      end
    end
    date ||= Time.now.iso8601

    blocks = fetch_cached_page_blocks(page_id, last_edited_time)
    content = format_notion_blocks(blocks)

    {
      'title' => title,
      'date' => date,
      'html' => content,
      'path' => "/#{page_name.downcase}",
      'id' => page_id
    }
  end
end

def fetch_notion_database_pages(database_name)
  content_notion_url = ENV['CONTENT_NOTION_URL']
  raise "CONTENT_NOTION_URL not found in environment variables" if content_notion_url.nil? || content_notion_url.empty?

  content_page_id = extract_page_id_from_notion_url(content_notion_url)
  content_page = notion_client.page(page_id: content_page_id)
  content_page_last_edited = content_page['last_edited_time'] || content_page[:last_edited_time] || Time.now.iso8601

  database_id_key = "#{content_page_id}_#{database_name}_#{content_page_last_edited}"
  database_id = $cache.fetch(database_id_key) do
    find_database_id(content_page_id, database_name)
  end

  # Don't cache the database query results - we cache individual pages instead
  pages = []
  response = notion_client.database_query(
    database_id: database_id,
    sorts: [{ 'property' => 'date', 'direction' => 'descending' }],
    page_size: 100
  )
  pages.concat(response['results'] || [])

  while response['has_more'] && response['next_cursor']
    response = notion_client.database_query(
      database_id: database_id,
      sorts: [{ 'property' => 'date', 'direction' => 'descending' }],
      page_size: 100,
      start_cursor: response['next_cursor']
    )
    pages.concat(response['results'] || [])
  end

  pages
end

def load_entity_notion_database_schema(entity_name)
  schema_path = File.join(File.dirname(__FILE__), 'config', 'notion_database_schemas', "#{entity_name}.yaml")
  YAML.load_file(schema_path)
end

def fetch_entity_props(page_props, prop_schema)
  prop_names = prop_schema['names'] || []
  prop_type = prop_schema['type']
  extract_method = prop_schema['extract']
  default_value = prop_schema['default']

  prop_names.each do |name|
    prop = page_props[name]
    next unless prop
    next unless prop['type'] == prop_type

    case prop_type
    when 'date'
      date_value = prop['date']
      return date_value['start'] if date_value && extract_method == 'start'
    when 'checkbox'
      return prop['checkbox']
    when 'multi_select'
      multi_select_value = prop['multi_select']
      if multi_select_value && extract_method == 'names_sorted'
        return multi_select_value.map { |item| item['name'] }.sort
      end
    end
  end

  default_value
end

def process_entity_page(page, path_prefix, properties_schema)
  properties = page['properties'] || {}
  page_id = page['id']
  last_edited_time = page['last_edited_time'] || page[:last_edited_time]

  cache_key = "entity_#{page_id}_#{last_edited_time}"

  $cache.fetch(cache_key) do
    extracted_props = {}

    properties_schema.each do |prop_name, prop_schema|
      next unless prop_schema

      if prop_name == 'content' && prop_schema['required']
        blocks = fetch_cached_page_blocks(page_id, last_edited_time)
        return nil if blocks.nil? || blocks.empty?
        content = format_notion_blocks(blocks)
        extracted_props['content'] = content
      else
        value = fetch_entity_props(properties, prop_schema)
        if prop_schema['required'] && (value.nil? || (value.respond_to?(:empty?) && value.empty?))
          return nil
        end
        extracted_props[prop_name] = value
      end
    end

    title = nil
    properties.each do |prop_name, prop|
      if prop['type'] == 'title' || prop[:type] == 'title'
        title = prop['title']&.first&.[]('plain_text') if prop['title']&.any?
        break if title
      end
    end
    title ||= 'Untitled'

    title_slug = title != 'Untitled' ? title.downcase.gsub(/[^a-z0-9]+/, '-').gsub(/^-|-$/, '') : ''
    short_id = page_id.gsub(/-/, '')[0..7]
    slug = "#{short_id}-#{title_slug}".gsub(/-+$/, '')

    {
      'title' => title,
      'date' => extracted_props['date'],
      'html' => extracted_props['content'],
      'path' => "#{path_prefix}/#{slug}",
      'draft' => extracted_props['draft'],
      'tags' => extracted_props['tags'],
      'id' => page_id
    }
  end
end

def fetch_entities(entity_name)
  entity_schema = load_entity_notion_database_schema(entity_name)
  database_name = entity_schema['database_name']
  path_prefix = entity_schema['path_prefix']
  properties_schema = entity_schema['properties'] || {}

  pages = fetch_notion_database_pages(database_name)

  pages.filter_map do |page|
    process_entity_page(page, path_prefix, properties_schema)
  end
end

File: src/notion_utils.ts
import { Client } from '@notionhq/client';
import * as fs from 'fs';
import * as path from 'path';
import * as yaml from 'js-yaml';
import { Cache, NullCache } from './cache';
import { fetchNotionPageBlocks, formatNotionBlocks } from './notion_page_html_utils';
import { Content } from './entity';

export interface EntitySchema {
  name: string;
  itemProperties: Record<string, PropertySchema>;
}

export interface PropertySchema {
  names?: string[];
  type: string;
  extract?: string;
  default?: unknown;
  required?: boolean;
}

export interface NotionPage {
  id: string;
  properties: Record<string, unknown>;
  last_edited_time?: string;
  [key: string]: unknown;
}

interface NotionBlock {
  id: string;
  type: string;
  [key: string]: unknown;
}

let notionClient: Client | null = null;

export function getNotionClient(): Client {
  if (!notionClient) {
    const token = process.env.NOTION_TOKEN;
    if (!token) {
      throw new Error('NOTION_TOKEN not found in environment variables');
    }
    notionClient = new Client({ auth: token });
  }
  return notionClient;
}

export async function fetchCachedPageBlocks(
  pageId: string,
  lastEditedTime: string,
  options: { cache?: Cache } = {}
): Promise<unknown[]> {
  const cache = options.cache || new NullCache();
  const cacheKey = `page_${pageId}_${lastEditedTime}_blocks`;
  return cache.fetch(cacheKey, async () => {
    return fetchNotionPageBlocks(getNotionClient(), pageId);
  });
}

export function extractPageIdFromNotionUrl(url: string): string {
  const uuidMatch = url.match(/-([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})$/i);
  if (uuidMatch) {
    return uuidMatch[1];
  }

  const hexMatch = url.match(/-([a-f0-9]{32})$/i);
  if (hexMatch) {
    const id = hexMatch[1];
    return `${id.substring(0, 8)}-${id.substring(8, 12)}-${id.substring(12, 16)}-${id.substring(16, 20)}-${id.substring(20, 32)}`;
  }

  throw new Error(`Could not extract page ID from Notion URL: ${url}`);
}

export async function findDatabaseId(
  contentPageId: string,
  databaseName: string
): Promise<string> {
  const client = getNotionClient();
  const blocks = await fetchAllBlocks(client, contentPageId);

  for (const block of blocks) {
    if (block.type === 'child_database') {
      const title = block.child_database?.title || '';
      if (title.toLowerCase().includes(databaseName.toLowerCase())) {
        return block.id;
      }
    }
  }

  throw new Error(`${databaseName.charAt(0).toUpperCase() + databaseName.slice(1)} database not found in CONTENT_NOTION_URL page`);
}

export async function findNotionPageByTitle(
  contentPageId: string,
  pageTitle: string
): Promise<string> {
  const client = getNotionClient();
  const blocks = await fetchAllBlocks(client, contentPageId);

  for (const block of blocks) {
    if (block.type === 'child_page') {
      const title = block.child_page?.title || '';
      if (title.toLowerCase() === pageTitle.toLowerCase()) {
        return block.id;
      }
    }
  }

  throw new Error(`${pageTitle} page not found in CONTENT_NOTION_URL page`);
}

export async function fetchNotionPageByName(
  pageName: string,
  options: { cache?: Cache } = {}
): Promise<Content> {
  const cache = options.cache || new NullCache();
  const contentNotionUrl = process.env.CONTENT_NOTION_URL;
  if (!contentNotionUrl || contentNotionUrl.trim() === '') {
    throw new Error('CONTENT_NOTION_URL not found in environment variables');
  }

  const contentPageId = extractPageIdFromNotionUrl(contentNotionUrl);
  const pageId = await findNotionPageByTitle(contentPageId, pageName);

  const client = getNotionClient();
  const pageResponse = await client.pages.retrieve({ page_id: pageId });
  const page = pageResponse as NotionPage;

  const properties = page.properties || {};
  const lastEditedTime = page.last_edited_time || '';

  const cacheKey = `page_${pageName}_${lastEditedTime}`;
  return cache.fetch(cacheKey, async () => {
    let title: string | null = null;
    for (const propName in properties) {
      const notionProperty = properties[propName];
      if (notionProperty.type === 'title' || (notionProperty as any).type === 'title') {
        if (notionProperty.title && notionProperty.title.length > 0) {
          title = notionProperty.title[0].plain_text;
          break;
        }
      }
    }
    title = title || pageName.charAt(0).toUpperCase() + pageName.slice(1);

    const publishedDateProperty = (properties.date || properties.Date || properties.created_time || properties['Created Time']) as 
      { type?: string; date?: { start?: string }; created_time?: string } | undefined;
    let publishedDate: string;
    if (publishedDateProperty) {
      if (publishedDateProperty.type === 'date' && publishedDateProperty.date?.start) {
        publishedDate = publishedDateProperty.date.start;
      } else if (publishedDateProperty.type === 'created_time' && publishedDateProperty.created_time) {
        publishedDate = publishedDateProperty.created_time;
      } else {
        publishedDate = new Date().toISOString();
      }
    } else {
      publishedDate = new Date().toISOString();
    }

    const blocks = await fetchCachedPageBlocks(pageId, lastEditedTime, { cache });
    const content = await formatNotionBlocks(getNotionClient(), blocks);

    return {
      id: pageId,
      name: pageName.toLowerCase(),
      title: title,
      publishedDate: publishedDate,
      html: content,
      slug: pageName.toLowerCase()
    };
  });
}

export async function fetchNotionDatabasePages(
  databaseName: string,
  options: { cache?: Cache } = {}
): Promise<NotionPage[]> {
  const cache = options.cache || new NullCache();
  const contentNotionUrl = process.env.CONTENT_NOTION_URL;
  if (!contentNotionUrl || contentNotionUrl.trim() === '') {
    throw new Error('CONTENT_NOTION_URL not found in environment variables');
  }

  const contentPageId = extractPageIdFromNotionUrl(contentNotionUrl);
  const client = getNotionClient();
  const contentPage = await client.pages.retrieve({ page_id: contentPageId });
  const contentPageLastEdited = (contentPage as NotionPage).last_edited_time || new Date().toISOString();

  const databaseIdKey = `${contentPageId}_${databaseName}_${contentPageLastEdited}`;
  const databaseId = await cache.fetch(databaseIdKey, async () => {
    return findDatabaseId(contentPageId, databaseName);
  });

  return fetchAllDatabasePages(client, databaseId);
}

export function loadEntityNotionDatabaseSchema(entityName: string): EntitySchema {
  const schemaPath = path.join(__dirname, '..', 'config', 'notion_database_schemas', `${entityName}.yaml`);
  const fileContents = fs.readFileSync(schemaPath, 'utf8');
  return yaml.load(fileContents) as EntitySchema;
}

export function fetchEntityProps(pageProps: Record<string, unknown>, propSchema: PropertySchema): unknown {
  const propNames = propSchema.names || [];
  const propType = propSchema.type;
  const extractMethod = propSchema.extract;
  const defaultValue = propSchema.default;

  for (const name of propNames) {
    const notionProperty = pageProps[name];
    if (!notionProperty) continue;
    if (notionProperty.type !== propType) continue;

    switch (propType) {
      case 'date':
        const datePropertyValue = notionProperty.date as { start?: string } | undefined;
        if (datePropertyValue && extractMethod === 'start') {
          return datePropertyValue.start;
        }
        break;
      case 'checkbox':
        return notionProperty.checkbox;
      case 'multi_select':
        const multiSelectValue = notionProperty.multi_select as Array<{ name: string }>;
        if (multiSelectValue && extractMethod === 'names_sorted') {
          return multiSelectValue.map((tagOption) => tagOption.name).sort();
        }
        break;
    }
  }

  return defaultValue;
}

export async function fetchAllBlocks(client: Client, blockId: string): Promise<NotionBlock[]> {
  const blocks: NotionBlock[] = [];
  let response = await client.blocks.children.list({
    block_id: blockId,
    page_size: 100
  });
  blocks.push(...(response.results as NotionBlock[]));

  while (response.has_more && response.next_cursor) {
    response = await client.blocks.children.list({
      block_id: blockId,
      page_size: 100,
      start_cursor: response.next_cursor
    });
    blocks.push(...(response.results as NotionBlock[]));
  }

  return blocks;
}

export async function fetchAllDatabasePages(client: Client, databaseId: string): Promise<NotionPage[]> {
  const pages: NotionPage[] = [];
  let response = await client.databases.query({
    database_id: databaseId,
    sorts: [{ property: 'date', direction: 'descending' }],
    page_size: 100
  });
  pages.push(...(response.results as NotionPage[]));

  while (response.has_more && response.next_cursor) {
    response = await client.databases.query({
      database_id: databaseId,
      sorts: [{ property: 'date', direction: 'descending' }],
      page_size: 100,
      start_cursor: response.next_cursor
    });
    pages.push(...(response.results as NotionPage[]));
  }

  return pages;
}

export { fetchPosts, fetchChirps, transformNotionPageToPost, transformNotionPageToChirp, transformNotionPageToEntity, fetchEntities } from './entity-utils';
